File: 2023/day8/part2.js
function getInput() {
    const fs = require('fs') // fs is the File System module

    let data = fs.readFileSync("input.txt") // read input.txt, store into data variable synchronously
    data = data.toString().split("\n")
    if (data[data.length-1].length == 0) {
        data.pop()
    }

    data = data.map((e) => e.split(" = "))
    return data; //convert data to string and split by EOL
}

const inputData = getInput()

// From Geeksforgeeks
function gcd(a, b) { 
    for (let temp = b; b !== 0;) { 
        b = a % b; 
        a = temp; 
        temp = b; 
    } 
    return a; 
} 
  
function lcmFunction(a, b) { 
    const gcdValue = gcd(a, b); 
    return (a * b) / gcdValue; 
} 

function solution(input) {
    let instructions = input[0][0]
    let locations = input.slice(2).map((e) => e[0])
    let destinations = input.slice(2).map((e) => e[1].replace(/[\(\)]/g, '').split(", "))

    //console.log(instructions)

    let curr = locations.reduce((a,e,i) => {
        if (e[2] == "A") a.push(i);
        return a;
    }, []);

    let steps = 0

    let lcm = 1

    while (curr.length) {
        for (let i = 0; i < curr.length; i++) {
            if (locations[curr[i]][2] == "Z") {
                lcm = lcmFunction(lcm, steps)
                curr.splice(i, 1)
            }
        }
        


        if (instructions[steps % instructions.length] == 'L') {
            for (let i = 0; i < curr.length; i++) {
                curr[i] = locations.indexOf(destinations[curr[i]][0])
            } 
        }
        else {
            for (let i = 0; i < curr.length; i++) {
                curr[i] = locations.indexOf(destinations[curr[i]][1])
            } 
        }

        steps++
    }

    return lcm
}

console.log("Part 2")
console.log(solution(inputData))

File: 2023/day8/part2.py
import math

def getInput(filename = "input.txt"):
    with open(filename,"r") as f:
        filedata = f.readlines()
    
    instructions = filedata[0].strip()
    locationdata = filedata[2:]

    locations = {}
    for location in locationdata:
        x = location.split(" = ")
        locations[x[0]] = x[1][1:-2].split(", ")

    return instructions,locations

instructions, locations = getInput()

def solution(instructions,locations):

    curr = [x for x in locations.keys() if x[-1] == "A"]
    
    ste = 1

    for x in curr:
        steps = 0
        # print(x)
        while not x[-1] == "Z":
            x = locations[x][0] if instructions[steps % len(instructions)] == "L" else locations[x][1]
            # print(x + ": " + str(steps % len(instructions)) + " " + instructions[steps % len(instructions)])
            steps += 1
        # print(x)
        # print(steps)
        ste = math.lcm(ste,steps)


    return ste


print(solution(instructions,locations))